rfcs/NNNN-decorator-roadmap.md

```diff
@@ -0,0 +1,278 @@
+---
+Status: Accepted
+Champions: "@justinfagnani"
+PR: https://github.com/lit/rfcs/pull/20
+---
+
+# Decorator Roadmap
+
+This RFC describes how Lit can migrate to standard decorators with the smoothest upgrade path possible for developers.
+
+## Objective
+
+Migrate to using standard decorators as the only decorator implementation and unify Lit's syntax around decorators as the only way to declare reactive properties. When the migration is complete, Lit will require a decorator-supporting environment or toolchain.
+
+### Goals
+
+- Make Lit decorators work without transpilation
+- Make the migration as easy as possible on developers
+- Have a single implementation for decorators
+- Unify Lit's reactive property syntax
+- Simplify ReactiveElement's property handling code
+- Increase component performance due to more static class initialization
+- Decrease code size with simpler decorator implementations
+
+### Non-Goals
+
+- Change our decorators API drastically (ie, split `@property()` into `@input()`, `@output()`, `@attribute`, etc)
+
+## Motivation
+
+As [standard decorators](https://github.com/tc39/proposal-decorators) are starting to ship in [compilers](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-0.html#decorators) and soon in VMs, we need to prepare for migrating Lit to use them.
+
+Standard decorators will allow us to unify our surface syntax which is currently different in plain JS and compiled sources using TypeScript or Babel. This will let us use decorators as the one way of declaring reactive properties, instead of also offering the `static properties` feature. Removing `static properties` in turn lets us remove the infrastructure for dynamically creating reactive properties, simplifying ReactiveElement.
+
+## Detailed Design
+
+Detailed design of the new decorators should be covered in another RFC. This RFC focuses on:
+
+- That standard decorators become the only way to declare reactive properties
+- That breaking changes are made to remove support for dynamically adding reactive properties, since that will not be used by the new decorators
+- A plan so that developers can incrementally migrate to standard decorators
+
+### Audiences
+
+If we group our developers by how they relate to decorators, we have four distinct audiences:
+
+- JavaScript developers not yet using decorators
+- TypeScript developers using legacy experimental decorators
+- TypeScript developers not yet using decorators because of forward compatibility concerns
+- TypeScript developers stuck using legacy experimental decorators because of backward compatibility concerns
```

Comment:
Should second point specify the legacy babel decorators?
Would the fourth option now be `JavaScript/TypeScript` developers using standard decorators.

Reply:
I think my thinking was that "JavaScript developers using decorators with Babel" as a single group because during migration they _have_ to be using standard decorators. They have no other choice as legacy babel mode is removed.

Reply:
Fair! I like this! I think I was coming from this retrospectively. When this RFC was written and implemented, we still supported old Babel decorators. And then chose the break them as part of Stage 1.